Feature: Add foreign keys between `SimulationConfiguration`, `Run` and `Token`

antonykamp opened the pull request:
Fixes #224 

Add foreign keys from `Run` to `SimulationConfiguration` and `SimulationConfiguration` to `Token`.

## PR checklist

- [x] Acceptance criteria fulfilled
- [x] Additional features are tested
- [x] Docs (code, wiki & diagrams) updated
- [ ] Dev-branch has been merged into local branch to resolve conflicts
- [ ] Tests and linter have passed AFTER local merge
- [x] Another dev reviewed and approved

antonykamp committed:
refactor: move SimulationConfiguration in file

Move SimulationConfiguration in file to reference class Token later

antonykamp committed:
feat: add reference to Token from SimulationConfiguration

Add foreign key from simulation configuration to token

antonykamp committed:
test: add tests for new reference

Add tests for new reference from SimulationConfiguration to Token.

antonykamp committed:
feat: add reference to SimulationConfiguration from Run

Add foreign key from run to simulation configuration

antonykamp committed:
test: add tests for new reference

Add tests for new reference from Run to SimulationConfiguration

LucasDerReisende commented on tests/implementor/test_table_run.py:
```diff
@@ -37,20 +48,37 @@ class TestRunSuccessfulInit:
+    @pytest.fixture
+    def token(self):
+        token = Token(name="owner", permission="admin", hashedToken="hash")
+        token.save()
```
does there need to be a `.save()`, just curious which you added it

antonykamp replied:
You can insert a new entry to the database by using `Token.create()` or by initializing it first and saving it with `.save()` later 😇

LucasDerReisende replied:
Yes, but does it need to be saved in the database for this test? Just asking, because I did not in my tests and it also worked

antonykamp replied:
I thought it could cause an error, when I want to reference a Token from a SimulationConfiguration with foreign key (see the following test case) 

https://github.com/BP2022-AP1/bp2022-ap1/blob/6263c730a63dce9038242b96eb88eba1b9567441/tests/implementor/test_table_run.py#L21)

LucasDerReisende replied:
Nevermind, I though about something else, but does not seem to be a problem, so I'm good :)

antonykamp replied:
Great 🙌 Thanks for the question, tho 😄

LucasDerReisende reviewed (commented):
looks good, I'm just interest, why you used `.save()` in the tests

LucasDerReisende reviewed (approved):
Looks good 🔥  keep up the good work

antonykamp committed:
Merge branch 'dev' into 224-add-forgein-keys-between-simulationconfiguration-run-token

antonykamp committed:
fix: tests

Add reference from Run to SimulationConfiguration and Token to tests

antonykamp committed:
fix: lint

antonykamp committed:
Merge branch 'dev' into 224-add-forgein-keys-between-simulationconfiguration-run-token

SaturnHafen commented on tests/implementor/test_table_run.py:
```diff
@@ -37,20 +48,37 @@ class TestRunSuccessfulInit:
 
+    @pytest.fixture
+    def token(self):
+        token = Token(name="owner", permission="admin", hashedToken="hash")
```
Should we add validation to prevent not hashed tokens from being entered in the database?

antonykamp replied:
Is it possible to distinguish between a random string and a hash other than prepending `hash`? 🤔 

IMO it's not needed because we insert the tokens by ourselves. 🙃 We'll generate the token, save a hashed version in the database and return the original to the user.

SaturnHafen replied:
Well if you look at (for example) a sha256 hash, it will look consistent (64 hexadecimal chars). If one were to add salt & pepper to the hash, the format would get even more obvious.
In any case, setting a parameter called `hashedToken` to something obviously not hashed, seems a little off to me :slightly_smiling_face:

SaturnHafen reviewed (commented):
A little later but: Why do we need this "join"?
Is it to create an audit-trail to trace which token created which simulation run/configuration?

antonykamp committed:
fix: specify name in reused fixtures

To reuse a fixture in the same file we have to specify the name and rename the function to prevent linting error.

antonykamp committed:
Merge branch 'dev' into 224-add-forgein-keys-between-simulationconfiguration-run-token

antonykamp commented:
> A little later but 

It's never too late ;) Thanks for the hints! 🙃 

>  Why do we need this "join"? Is it to create an audit-trail to trace which token created which simulation run/configuration?

Exactly! Otherwise, we couldn't tell which simulation configuration belongs to which user or which run belongs to which simulation configuration.

antonykamp committed:
Merge branch 'dev' into 224-add-forgein-keys-between-simulationconfiguration-run-token
